tests_integ/helpers/base_test.py

```diff
@@ -0,0 +1,379 @@
+import logging
+import os
+
+try:
+    from pathlib import Path
+except ImportError:
+    from pathlib2 import Path
+from unittest.case import TestCase
+
+import boto3
+import pytest
+import yaml
+from botocore.exceptions import ClientError
+from botocore.config import Config
+from tests_integ.helpers.deployer.deployer import Deployer
+from tests_integ.helpers.helpers import transform_template, verify_stack_resources, generate_suffix, create_bucket
+from tests_integ.helpers.file_resources import FILE_TO_S3_URI_MAP, CODE_KEY_TO_FILE_MAP
+
+LOG = logging.getLogger(__name__)
+STACK_NAME_PREFIX = "sam-integ-stack-"
+S3_BUCKET_PREFIX = "sam-integ-bucket-"
+
+
+class BaseTest(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.tests_integ_dir = Path(__file__).resolve().parents[1]
+        cls.resources_dir = Path(cls.tests_integ_dir, "resources")
+        cls.template_dir = Path(cls.resources_dir, "templates", "single")
+        cls.output_dir = Path(cls.tests_integ_dir, "tmp")
+        cls.expected_dir = Path(cls.resources_dir, "expected", "single")
+        cls.code_dir = Path(cls.resources_dir, "code")
+        cls.s3_bucket_name = S3_BUCKET_PREFIX + generate_suffix()
+        cls.session = boto3.session.Session()
+        cls.my_region = cls.session.region_name
+        cls.s3_client = boto3.client("s3")
+        cls.api_client = boto3.client("apigateway")
+        cls.lambda_client = boto3.client("lambda")
+        cls.iam_client = boto3.client("iam")
+        cls.api_v2_client = boto3.client("apigatewayv2")
+        cls.sfn_client = boto3.client("stepfunctions")
+
+        cls.file_to_s3_uri_map = FILE_TO_S3_URI_MAP
+        cls.code_key_to_file = CODE_KEY_TO_FILE_MAP
+
+        if not cls.output_dir.exists():
+            os.mkdir(str(cls.output_dir))
+
+        cls._upload_resources(FILE_TO_S3_URI_MAP)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls._clean_bucket()
+
+    @classmethod
+    def _clean_bucket(cls):
+        """
+        Empties and deletes the bucket used for the tests
+        """
+        s3 = boto3.resource("s3")
+        bucket = s3.Bucket(cls.s3_bucket_name)
+        object_summary_iterator = bucket.objects.all()
+
+        for object_summary in object_summary_iterator:
+            try:
+                cls.s3_client.delete_object(Key=object_summary.key, Bucket=cls.s3_bucket_name)
+            except ClientError as e:
+                LOG.error(
```

Comment:
Should this cause a test to fail? If there is data left around, will that cause issue for the next run of the tests?

Reply:
I'm not sure cleanup of a test is part of the test itself, so not sure it should fail.
What we could do is attempt to clean it again and if it fails set this resource name aside and attempt to clean it again at the end of the run.

This shouldn't cause an issue for the next run as we are adding a simple random suffix to each of the objects, stacks and buckets, that we create (see `generate_suffix()`). I also added this suffix so that we could run the tests in parallel but for now it's not an option as we get throttled.